spec/reactors/consensus.md

```diff
@@ -0,0 +1,261 @@
+	BlockPartSetHeader types.PartSetHeader
+	BlockParts         *bits.BitArray
+	IsCommit           bool
+}
+```
+
+Upon receiving this message, the node will only modify the peer's round state under these conditions:
+- The `Height` specified in the message aligns with the peer's current `Height`.
+- The `Round` matches the most recent round known for the peer OR the message indicates the block's commitment i.e., `IsCommit` is `ture`.
+
+Following these verifications, the node will then update its peer state's `ProposaBlockPartSetHeader` and `ProposaBlockParts` based on the `BlockPartSetHeader` and `BlockParts` values from the received message.
+
+[//]: # (The BlockParts field in the message appears to represent the parts of a specific proposal that the sending peer has received from all its connections. This informs the receiving peer about the parts the sending peer possesses, possibly reducing the number of block parts sent subsequently.)
+
+[//]: # (Question: Does this message also signify that the sender has the entire proposal? 
+or can a node send this merely based on the observed votes?  
+Answer: After further investigation, it looks like that this is purely based on votes and does not signify the proposal completeion and the sender side. -->)
+
+## Network Traffic Analysis
+
+The following section provides the analysis of the network traffic generated by the `DataChannel` protocol.
+Essentially, the focus is on the [`BlockPart`](#block-part) message as well as the [`Proposal`](#proposal), which are the most frequently transmitted messages in the `DataChannel` protocol.
+
+We denote `block_part_size` as the size of a block part in bytes and `proposal_header_size` as the size of a proposal header in bytes.
+Suppose `proposal(H,R)` denotes the proposal at height and round `H` and `R`, respectively.
+With this notation we assume there is only one valid proposal in the network for a given round and height.
+`proposal(H,R).total` denotes the number of parts in that proposal.
+
+For every block proposal, peers start gossiping their obtained proposal and constituent block parts to their connected peers.
+Both sending and receiving ends monitor the block parts they've exchanged with their counterpart (either dispatched to or received from) and mark it as seen by the other peer.
+Peers keep exchanging block parts of a proposal until either 1) all block parts of the intended proposal have been successfully transmitted between the two peers or 2) one of the peer's round state updates (and points to a new height and round with a different proposal).
+In the latter case, the peer whose state has advanced still sends block parts to the other peer until all the parts are transmitted or until the receiving peer's round state is also updated.
+
+Worst Case Scenario: The worst case occurs when both peers coincidentally choose the same block part index at the same moment and initiate the sending process concurrently, a scenario that's rather unlikely.
```

Comment:
I'm not sure if this is actually what we are seeing in the testnet experiments, but we often see that some heights the block parts are not distributed randomly, which I think results in a worst case scenario more often than what we would expect. 
![normalized_receipt_of_block_parts_with_completion_times_per_height](https://github.com/celestiaorg/celestia-core/assets/42654277/a0836404-b39e-4413-8b81-81e9f15ff11b)

Reply:
to expand on this slightly, an example height that is less random and more in order are 30 or 31. Where heights such as 33 or 20 are randomly distributed relative to the peer that was tracing.

Reply:
Thanks for sharing this! I can see that the higher indices are mostly chosen close to the end of block completion. This might not only be due to a non-random selection process but also could be influenced by the sequence in which parts of the block proposal are released to the network. I guess based on this figure, the initial indices are released first and that could be why we see the first indices are circulated earlier. 

If we were to verify whether this figure indicates the worst case scenraio (based on the doc), we would need to examine the index of the individual block parts communicated per p2p connections and see if more than one instance of each block part shows up or not. (I am going to extract this, or see if we have enough metrics to verify this)